src/handlers.cpp

```diff
@@ -2437,85 +2467,153 @@ void PushProfileTask::on_get_impus_failure(CassandraStore::Operation* op,
-    }
-    else
-    {
-      event.add_var_param("Charging addresses unchanged");
-    }
+  Cache::PutRegData* put_reg_data =
+    _cfg->cache->create_PutRegData(_impus,
+                                   default_public_id,
+                                   Cache::generate_timestamp(),
+                                   _cfg->record_ttl);
+  SAS::Event event(this->trail(), SASEvent::CACHE_PUT_REG_DATA, 0);
 
-    CassandraStore::Transaction* tsx =
-      new CacheTransaction(this,
-                           &PushProfileTask::update_reg_data_success,
-                           &PushProfileTask::update_reg_data_failure);
-    CassandraStore::Operation*& op = (CassandraStore::Operation*&)put_reg_data;
-    _cfg->cache->do_async(op, tsx);
+  std::string impus_str = boost::algorithm::join(_impus, ", ");
+  event.add_var_param(impus_str);
 
-    SAS::report_event(event);
+  if (_ims_sub_present)
+  {
+    TRC_INFO("Updating IMS subscription from PPR");
+    put_reg_data->with_xml(_ims_subscription);
+    event.add_compressed_param(_ims_subscription, &SASEvent::PROFILE_SERVICE_PROFILE);
+  }
+  else
+  {
+    event.add_compressed_param("IMS subscription unchanged", &SASEvent::PROFILE_SERVICE_PROFILE);
+  }
+
+  event.add_static_param(RegistrationState::UNCHANGED);
+  event.add_var_param("");
 
+  if (_charging_addrs_present)
+  {
+    TRC_INFO("Updating charging addresses from PPR");
+    event.add_var_param(_charging_addrs.log_string());
+    put_reg_data->with_charging_addrs(_charging_addrs);
   }
   else
   {
-    send_ppa(DIAMETER_REQ_SUCCESS);
+    event.add_var_param("Charging addresses unchanged");
   }
+
+  CassandraStore::Transaction* tsx =
+    new CacheTransaction(this,
+                         &PushProfileTask::update_reg_data_success,
+                         &PushProfileTask::update_reg_data_failure);
+  CassandraStore::Operation*& op = (CassandraStore::Operation*&)put_reg_data;
+  _cfg->cache->do_async(op, tsx);
+
+  SAS::report_event(event);
+}
+
+void PushProfileTask::on_get_primary_impus_failure(CassandraStore::Operation* op,
+                                                   CassandraStore::ResultCode error,
+                                                   std::string& text)
+{
+  SAS::Event event(this->trail(), SASEvent::CACHE_GET_ASSOC_DEF_IMPU_FAIL, 0);
+  SAS::report_event(event);
+  TRC_DEBUG("Cache query failed with rc %d", error);
+  send_ppa(DIAMETER_REQ_FAILURE);
+}
+
+void PushProfileTask::on_get_primary_impus_success(CassandraStore::Operation* op)
+{
+  Cache::GetAssociatedPrimaryPublicIDs* get_primary_public_ids =
+    (Cache::GetAssociatedPrimaryPublicIDs*)op;
+  get_primary_public_ids->get_result(_default_impus);
+
+  if (_default_impus.empty())
+  {
+    TRC_INFO("No cached default public IDs found for private ID %s - failed to "
+             "update charging addresses", _impi.c_str());
+    SAS::Event event(this->trail(), SASEvent::CACHE_GET_ASSOC_DEF_IMPU_FAIL, 0);
```

Comment:
You're using this log for the second time so you should increment the instance ID from 0 to 1.